Fix cmake config file generation to use correct function

9prady9 opened the pull request:
Earlier to this change, the codebase was using regular configure_file
to generate the package config file which renders the install files
non-relocatable. This commit refactors the regular configure_file to
configure_package_config_file command.

Apart from the major change mentioned above, this commit also did a couple
of additional changes.
- adds openmp flags to the interface link options of hipSYCL-rt target so
  that this flag is passed to any application using cmake target. This was required on Arch Linux, without which examples are failing to build with `undefined reference` errors.
- Fixed the include path values passed to hipSYCL-rt target

9prady9 commented:
I have question about `examples` project ? Are they supposed to be installed along with libs? Because that doesn't seem to be happening right now.

9prady9 commented:
@illuhad I also had to add `stdc++` as linker flag for examples while using `cuda` platform, which isn't usually the case in most projects I have worked on, with CMake especially. What I am missing in `src/runtime/CMakeLists.txt` that is causing this. Any idea ?

#### Update
Alright, so using `clang++` instead of `clang` for CLANG variable resolved the problem with linking. However, I am getting all zeros in vector add example.
```
./vector_add/vector_add 
Running on GeForce GTX 1060 3GB
Result: 
0
0
0
0
0
```

illuhad commented:
Thank you.

>  This was required on Arch Linux, without which examples are failing to build with undefined reference errors.

I'm also on Arch Linux and cannot reproduce the issue. Please describe how you are compiling the application. `syclcc` already automatically links against OpenMP when targeting CPU, so this should not be needed.
Overall it's very dangerous to just hardcode OpenMP stuff because OpenMP flags are incompatible with clang CUDA due to a clang bug, so this might break CUDA toolchains.

> I have question about examples project ? Are they supposed to be installed along with libs? Because that doesn't seem to be happening right now.

No - the examples project was more of a toy, but it hasn't been touched in quite some time.

> However, I am getting all zeros in vector add example.

Please post the exact command line that you have used compile hipSYCL and your application.

@psalz @fknorr This touches the CMake integration, maybe you want to have a look at it?

9prady9 commented:
Thanks for responding promptly.

> I'm also on Arch Linux and cannot reproduce the issue. Please describe how you are compiling the application.

After building the library, I did the following:

1. Created `build` under examples since these seem to be indepedent of main project
2. Did `cmake .. -DhipSYCL_DIR:PATH=<install_path>/lib/cmake/hipSYCL`
3. Step 2 finishes fine once all cmake options are set correctly.
4. While `make` I get the undefined reference errors for OpenMP symbol.

> syclcc already automatically links against OpenMP when targeting CPU, so this should not be needed.

I understand now why this didn't seem to happen earlier. It is due to the same reason which was causing the error `error adding symbols: DSO missing from command line` when I had `clang` instead of `clang++`.

> Overall it's very dangerous to just hardcode OpenMP stuff because OpenMP flags are incompatible with clang CUDA due to a clang bug, so this might break CUDA toolchains.

I wasn't using hardcoded flags, rather the cmake variable provided by the cmake OpenMP Module/config file. Neverthless, I am not OpenMP expert, I wouldn't know how flag interplay works between clang, CUDA

I should have noticed the `clang` instead of `clang++` -  would have saved some time. I wonder if there is a cmake way to check if the user provided clang++ vs clang and warn/error the user/developer if it is not clang++. That would be nice.

Irrespective of the fixing the above clang++ problem I had earlier, the output of vector add with `cuda` platform is still all zeros. I tried it using simple command line(shown below) approach and cmake project approach also, just to be sure.
```
../../build/pkg/bin/syclcc vector_add.cpp -o a.out --hipsycl-platform=cuda --hipsycl-gpu-arch=sm_61
```

@illuhad I corrected OpenMP flag addition, please check. With OpenMP flag variable gone, this change is purely CMake config file related only.

illuhad commented:
Ah, good to know that using `clang` instead of `clang++` also breaks OpenMP.

> I wonder if there is a cmake way to check if the user provided clang++ vs clang and warn/error the user/developer if it is not clang++. That would be nice.

Usually clang++ is called `clang++` or `clang++-VERSION`. We could probably check if the argument matches that pattern, but I don't think that's very reliable - there are clang distributions that have other names.

> Irrespective of the fixing the above clang++ problem I had earlier, the output of vector add with cuda platform is still all zeros. I tried it using simple command line(shown below) approach and cmake project approach also, just to be sure.

Regarding the CUDA issue:
Can you please post the output of
* your `syclcc` invocation when setting the environment variable `HIPSYCL_DRYRUN=1` (this will print the clang invocation instead of executing it)
* your application when setting the environment variable `HIPSYCL_DEBUG_LEVEL=3`

EDIT: Are you using this code https://github.com/illuhad/hipSYCL/blob/develop/doc/examples.md ? If you have made any changes, please also post the code.

fodinabor commented:
hmm.. that really should **not** make a difference, all those are symlinks to clang-VERSION (at least on clang 10+, don't have anything below that to check..)
![image](https://user-images.githubusercontent.com/5982050/96720838-94b29900-13ab-11eb-8ba0-5e3972cb9266.png)

just saying, that there should be another root cause and trying to match clang++ only, would not make to much sense, I think..

illuhad commented:
@fodinabor It definitely makes a difference - `clang++` for example links with the C++ standard library which `clang` definitely does not. How they do that when they are just symlinks I don't know. Maybe clang checks if it is invoked as `clang++` or `clang` using `argv[0]` or something like that...

9prady9 commented:
> your syclcc invocation when setting the environment variable HIPSYCL_DRYRUN=1 (this will print the clang invocation instead of executing it)

```bash
HIPSYCL_DRYRUN=1 ../../build/pkg/bin/syclcc vector_add.cpp -o a.out --hipsycl-platform=cuda --hipsycl-gpu-arch=sm_61
syclcc warning: No optimization flag was given, optimizations are disabled by default. Performance may be degraded. Compile with e.g. -O2/-O3 to enable optimizations.
/usr/bin/clang++ -std=c++17 -fplugin=/home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/bin/../lib/libhipSYCL_clang.so -D__HIPSYCL_CLANG__ -D__HIPSYCL__ -I/home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/bin/../include/ -x cuda --cuda-gpu-arch=sm_61 --cuda-path=/opt/cuda-10.1 -D__HIPSYCL_ENABLE_CUDA_TARGET__ -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -rpath /opt/cuda-10.1/lib64 -L/opt/cuda-10.1/lib64 -lcudart -rpath /home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/bin/../lib/ -L/home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/bin/../lib/ -lhipSYCL-rt vector_add.cpp -o a.out
```

Debug Level 3 output is given below
```bash
HIPSYCL_DEBUG_LEVEL=3 LD_LIBRARY_PATH=/home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/lib:/opt/cuda-10.1/lib64 ./a.out 
[hipSYCL Info] dag_manager: DAG manager is alive!
[hipSYCL Info] backend_manager: Registering CUDA backend...
[hipSYCL Info] backend_manager: Registering OpenMP backend...
[hipSYCL Info] runtime: ******* rt launch initiated ********
Running on GeForce GTX 1060 3GB
[hipSYCL Info] data_region: constructed with page table dimensions 1 1 1
[hipSYCL Info] data_region: constructed with page table dimensions 1 1 1
[hipSYCL Info] data_region: constructed with page table dimensions 1 1 1
[hipSYCL Info] dag_manager: Checking DAG flush opportunity...
[hipSYCL Info] dag_manager: Submitting asynchronous flush...
[hipSYCL Info] buffer_impl::~buffer_impl: Preparing submission of writeback...
[hipSYCL Info] [hipSYCL Info] dag_manager: Checking DAG flush opportunity...
[hipSYCL Info] dag_manager: Submitting asynchronous flush...
[hipSYCL Info] buffer_impl::~buffer_impl: Waiting for operations to complete...
dag_manager [async]: Flushing![hipSYCL Info] buffer_impl::~buffer_impl: dag node is registered as user but not marked as submitted, performing emergency DAG flush.
[hipSYCL Info] dag_manager: Submitting asynchronous flush...
[hipSYCL Info] dag_manager: waiting for async worker...

[hipSYCL Info] dag_builder: DAG contains operation: kernel: ZZ3addRN7hipsycl4sycl5queueERKSt6vectorIfSaIfEES7_ENKUlRNS0_7handlerEE_clES9_EUlNS0_2idILi1EEEE_
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info]     --> requires: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info]     --> requires: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info]     --> requires: MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] dag_builder: DAG contains operation: MEM_REQ: R host_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info]     --> requires: kernel: ZZ3addRN7hipsycl4sycl5queueERKSt6vectorIfSaIfEES7_ENKUlRNS0_7handlerEE_clES9_EUlNS0_2idILi1EEEE_
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] dag_builder: DAG contains operation: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] dag_builder: DAG contains operation: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] dag_builder: DAG contains operation: MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] dag_manager [async]: Releasing dead users of data region 0x563bad835f70
[hipSYCL Info] dag_manager [async]: Releasing dead users of data region 0x563bad8361a0
[hipSYCL Info] dag_manager [async]: Releasing dead users of data region 0x563bad8363d0
[hipSYCL Info] dag_manager [async]: Submitting node to direct scheduler!
[hipSYCL Info] dag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Warning] dag_direct_scheduler: Detected a requirement that is neither of discard access mode (SYCL 1.2.1) nor noinit property (SYCL 2020) that accesses uninitialized data. Consider changing to discard/noinit. Optimizing potential data transfers away.
[hipSYCL Info] dag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Warning] dag_direct_scheduler: Detected a requirement that is neither of discard access mode (SYCL 1.2.1) nor noinit property (SYCL 2020) that accesses uninitialized data. Consider changing to discard/noinit. Optimizing potential data transfers away.
[hipSYCL Info] dag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] multi_queue_executor: Processing node 0x563bad836670[hipSYCL Info] multi_queue_executor: Dispatching to lane 0: Memcpy: CPU-Device0 #4 {0, 0, 0}+{1, 1, 5}-->CUDA-Device0 #4 {0, 0, 0}+{1, 1, 5}{1, 1, 5}
[hipSYCL Info] multi_queue_executor: Processing node 0x563bad836950[hipSYCL Info] multi_queue_executor: Dispatching to lane 2: kernel: ZZ3addRN7hipsycl4sycl5queueERKSt6vectorIfSaIfEES7_ENKUlRNS0_7handlerEE_clES9_EUlNS0_2idILi1EEEE_
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 5} #4
   MEM_REQ: W global_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] deferred_pointer: Initialized to 0x7ffb96200400
[hipSYCL Info] deferred_pointer: Initialized to 0x7ffb96200000
[hipSYCL Info] deferred_pointer: Initialized to 0x7ffb96200200
[hipSYCL Info] dag_manager [async]: Submitting node to direct scheduler!
[hipSYCL Info] dag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: R host_buffer {0, 0, 0}+{1, 1, 5} #4
[hipSYCL Info] multi_queue_executor: Processing node 0x563bad836b80[hipSYCL Info]  --> Synchronizes with other queue for node: 0x563bad836950 lane = 2
[hipSYCL Info] multi_queue_executor: Dispatching to lane 0: Memcpy: CUDA-Device0 #4 {0, 0, 0}+{1, 1, 5}-->CPU-Device0 #4 {0, 0, 0}+{1, 1, 5}{1, 1, 5}
[hipSYCL Info] dag_manager [async]: Nothing to do
[hipSYCL Info] dag_manager [async]: Nothing to do
[hipSYCL Info] buffer_impl::~buffer_impl: Waiting for operations to complete...
[hipSYCL Info] buffer_impl::~buffer_impl: Waiting for operations to complete...
Result: 
0
0
0
0
0
```

> Are you using this code https://github.com/illuhad/hipSYCL/blob/develop/doc/examples.md ? If you have made any changes, please also post the code.

No changes, just running it as it is.

9prady9 commented:
> @fodinabor It definitely makes a difference - `clang++` for example links with the C++ standard library which `clang` definitely does not. How they do that when they are just symlinks I don't know. Maybe clang checks if it is invoked as `clang++` or `clang` using `argv[0]` or something like that...

Clang hidden options does show driver mode option.
```
  --driver-mode=<value> Set the driver mode to either 'gcc', 'g++', 'cpp', or 'cl'
```
So, it seems like clang driver definitely does what @illuhad suggested. Get the name of clang/clang++ file running and then switch to appropriate driver mode.

illuhad commented:
@9prady9 thanks for the output, I can reproduce. Not yet sure what happens here - the kernel as well as data transfers seem to execute correctly.

EDIT: Issue might be initialization of buffers with const pointers specifically.

9prady9 commented:
> @9prady9 thanks for the output, I can reproduce. Not yet sure what happens here - the kernel as well as data transfers seem to execute correctly.
> 
> EDIT: Issue might be initialization of buffers with const pointers specifically.

Phew, that is it. As soon as I removed the const qualifier on `std::vector`s passed to the function, it works as expected. Hm, interesting

9prady9 commented:
Having `const` is valid and correct, but surprisingly it isn't working. If I only remove `const` for `a` vector, then the output is vector `b`. So, it seems like if we pass `const` vector, it is being ignored and another buffer(memory allocation) is being used somehow.

illuhad commented:
@9prady9 The issue is that buffers constructed from const pointers are not correctly initialized. For non-const pointers it works as expected. I have narrowed down the cause to an optimization gone wrong: When hipSYCL thinks that a buffer has not yet been initialized it will not transfer data to device (why copy unitialized data?). But the mechanics for const initialization are different such that in this case it doesn't actually realize that the buffer has been initialized, and it optimizes the data transfer away.

9prady9 commented:
I am curious about where this is happening in the code, can you please share the file and line info. Thanks!

Also, I have been wondering, where do I ask simple questions that aren't issues. Do you guys have some forum or shall I just open an issue on GitHub ?

illuhad commented:
@9prady9 Here you can see that it first initializes the buffer, and then copies:
https://github.com/illuhad/hipSYCL/blob/0e8fe1dc109b20a95c84cf66a8f03d307a674720/include/hipSYCL/sycl/buffer.hpp#L313
`copy_host_content()` effectively just does a memcpy on the internal buffer storage. That means that the content has changed without the runtime (the `rt::data_region` member) being informed that it now contains valid data.

We don't have a forum, so currently you can either drop me an email or open an issue. But maybe we should introduce something like gitter or so - until now nobody has asked for that ;)

9prady9 commented:
Thank you. I did reach those lines and then saw memcpy and was wondering what I am missing. I understand now, cool.

That is alright, I can open an issue for any questions I have until you guys bring up something like that.

The one I have right now is the following:

Is there a way to spit out PTX when using CUDA platform or whatever IR that the code is getting generated for the sycl code user has in the source file ? I checked the options of `syclcc` and it doesn't seem like it has one.

illuhad commented:
> Is there a way to spit out PTX when using CUDA platform or whatever IR that the code is getting generated for the sycl code user has in the source file ? I checked the options of syclcc and it doesn't seem like it has one.

Since hipSYCL effectively just invokes `clang -x cuda` or `clang -x hip`, you want to check the clang documentation :) All flags that work with clang CUDA/HIP also work with hipSYCL. All flags that you give `syclcc` that are not specific to hipSYCL will just be forwarded to the underlying compiler, so you could e.g. just do `syclcc --hipsycl-platform=cuda --hipsycl-gpu-arch=... -emit-llvm -S ...` to dump LLVM IR.

psalz commented on cmake/hipsycl-config.cmake.in:
```diff
@@ -48,7 +51,14 @@ if(HIPSYCL_ROCM_BACKEND_AVAILABLE)
+
+string(REPLACE "|" ";" HIPSYCL_PLATFORM_OPTIONS ${HIPSYCL_PLATFORMS_STRING})
+# Add canonical platform aliases to platform options list
+list(APPEND HIPSYCL_PLATFORM_OPTIONS "host" "hipcpu" "nvidia" "amd" "hip" "hcc")
```
This list includes all potential options, even if those backends aren't available (as opposed to `HIPSYCL_PLATFORMS_STRING`, which only includes available backends). Besides, I don't think we should be advertising the non-canonical names, these exist mostly for backwards compatibility (and should be removed at some point imo).

That being said, I realize now that the whole mechanism of building ` HIPSYCL_PLATFORMS_STRING` is needlessly convoluted, as a simple list append followed by a `list(JOIN ...)`  would achieve the same thing. While you're at it, maybe you could change that..? :)

9prady9 replied:
> This list includes all potential options, even if those backends aren't available (as opposed to HIPSYCL_PLATFORMS_STRING, which only includes available backends). Besides, I don't think we should be advertising the non-canonical names, these exist mostly for backwards compatibility (and should be removed at some point imo).

I have modified the `HIPSYCL_PLATFORM` to accept values from a given list. I wasn't sure if the values from [lines 70-78][1] are required or not. Given that they are essentially aliases to the three options `cuda`, `rocm` and `cpu`, in case backwards compatibility is the reason they are present, I added them to the options list as well. Note that the default value of this variable is still "" empty so that environment variables can be used.

I can revert the change to `HIPSYCL_PLATFORM` to being a regular string where the user has to type in the value if that is preferred over the cycling through a list. Please let me know.

About the removing the aliases at some point, those options can be easily removed by deleting the below line.

```cmake
# Add canonical platform aliases to platform options list
list(APPEND HIPSYCL_PLATFORM_OPTIONS "host" "hipcpu" "nvidia" "amd" "hip" "hcc")
```

> That being said, I realize now that the whole mechanism of building HIPSYCL_PLATFORMS_STRING is needlessly convoluted, as a simple list append followed by a list(JOIN ...) would achieve the same thing. While you're at it, maybe you could change that..? :)

I agree too. The reason I left it out is the following:
1. Right now, `HIPSYCL_PLATFORMS_STRING` is ending up being something like ` cpu | cuda | rocm ` which is nice to read in doc string for that particular CMake option instead of  semi-colon separated list.
2. `HIPSYCL_PLATFORMS_STRING` can easily be converted to a CMake option/string list which can be used as option list for `HIPSYCL_PLATFORM`.

[1]: https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L70-L78

psalz replied:
> I can revert the change to HIPSYCL_PLATFORM to being a regular string where the user has to type in the value if that is preferred over the cycling through a list. Please let me know.

I don't use the CMake GUI, but it is my understanding that the `STRINGS` property on cache entries is merely a visual aid for the GUI, and doesn't perform any form of validation. I think it should thus (1) not offer the option of selecting platforms that aren't available, and (2) not show any aliases, giving the impression that there is any difference between them, if in fact there is not (and we don't want to encourage their usage in the first place). Bottom line is, I think showing the possible options in the GUI is a good UX improvement, only it should be the right set of options :).

> I agree too. The reason I left it out is the following:

Yes, what I meant is that we could go about it in the opposite direction: Instead of building that pipe-separated string with the cumbersome `if() / else()` logic, first the `HIPSYCL_PLATFORM_OPTIONS` list could be built ([here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L23), [here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L34) and [here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L45)) and then converted into the nice looking string using `list(JOIN HIPSYCL_PLATFORM_OPTIONS " | " HIPSYCL_PLATFORMS_STRING)`.

9prady9 replied:
> Point 1

It is not cmake GUI alone. CMake curses GUI i.e. the command line will also use this list of STRINGS. It doesn't do any validation, however it does help users using CMake curses GUI to choose the right option without the need to type anything. Okay, only show `cpu | cuda | rocm`, if the user needs to, they will enter the other values via environment variable. Shall I proceed for that change ?

> Point 2

Sure, I can look into changing that logic.

psalz replied:
> Okay, only show cpu | cuda | rocm, if the user needs to, they will enter the other values via environment variable. Shall I proceed for that change ?

Only show whatever is available, i.e., the contents of `HIPSYCL_PLATFORMS_STRING` (or rather, `HIPSYCL_PLATFORM_OPTIONS`, if you do that refactoring). They will never need to provide any other value.

> Sure, I can look into changing that logic.

Great, thank you!

psalz commented on cmake/hipsycl-config.cmake.in:
```diff
@@ -73,8 +83,6 @@ elseif(HIPSYCL_PLATFORM MATCHES "cuda|nvidia")
 elseif(HIPSYCL_PLATFORM MATCHES "rocm|amd|hip|hcc")
   set(HIPSYCL_PLATFORM_CANONICAL "rocm")
-else()
-  message(SEND_ERROR "Unknown hipSYCL platform '${HIPSYCL_PLATFORM}'")
```
Why remove this?

9prady9 replied:
nice catch.

With the new logic, either the user leaves `HIPSYCL_PLATFORM` empty or sets one of the values by cycling down the list of options. So, I thought it is not possible to have any other value. But I now get it that, the environment variable can be a wrong string. I will revert this removal.

psalz commented:
Cool, I didn't even know `configure_package_config_file` existed. One never stops learning with CMake (not really a desirable quality for a build system though...).

> But maybe we should introduce something like gitter or so - until now nobody has asked for that ;)

@illuhad You could also try to get into the GitHub Discussions beta!

9prady9 commented:
> GitHub Discussions

That is interesting, I didn't know about GitHub Discussions.

9prady9 commented:
@psalz Please have a look at it now, I have pushed some changed based on our earlier discussions.

psalz commented:
Sorry for the delay. Also apparently I commented on the [commit](https://github.com/illuhad/hipSYCL/commit/706ee576b5fac57d8dfb77b1041451de3b36365c) instead of starting a review...

There is also #362, which has some conflicts with this. Also it seems like `HIPSYCL_PLATFORM` and `HIPSYCL_GPU_ARCH` are being deprecated, so we should probably reflect this in the help string. Maybe the changes in here regarding `HIPSYCL_PLATFORM` should be moved into a separate PR altogether..?

9prady9 commented:
> Sorry for the delay. Also apparently I commented on the [commit](https://github.com/illuhad/hipSYCL/commit/706ee576b5fac57d8dfb77b1041451de3b36365c) instead of starting a review...
> 
> There is also #362, which has some conflicts with this. Also it seems like `HIPSYCL_PLATFORM` and `HIPSYCL_GPU_ARCH` are being deprecated, so we should probably reflect this in the help string. Maybe the changes in here regarding `HIPSYCL_PLATFORM` should be moved into a separate PR altogether..?

Perhaps, I can rebase it of `develop` branch after #362 is merged in and do the necessary changes ?

9prady9 commented:
> I'm not sure I understand. In the previous behavior, if `HIPSYCL_PLATFORM` wasn't set, it would first try to use `HIPSYCL_PLATFORM_ENV`. If that wasn't set either, and if only a single backend was available anyway, it would pick that one. Otherwise it would throw an error. In the new behavior, if `HIPSYCL_PLATFORM` and `HIPSYCL_PLATFORM_ENV` aren't set, it will always throw an error, regardless of the number of available backends, correct?

True, my bad in misinterpreting `>` as `>=`, I fixed it now.

9prady9 commented:
> `JOIN` builds a string out of a list by gluing the items together with a given _glue_. But again, no big deal, as this is being deprecated anyway, we shouldn't be spending too much time on it...

never knew the glue part :) good to know, thanks

9prady9 commented:
Any ETA on when #370 will be finished ?

illuhad commented:
@psath Can you comment on the state of #370? I see you have added some commits recently.

psath commented:
> @psath Can you comment on the state of #370? I see you have added some commits recently.

#370 Is ready to go.

The new commits are minor fixes. When I was hastily removing the canonicalization of `HIPSYCL_TARGETS`, I inadvertently removed the line that actually set `--hipsycl-targets`. Just reintroducing it, then adjusting ROCm path and host CXX specification to require the `hip` and `omp` targets respectively.

illuhad commented:
#370 is in, can you please resolve merge conflicts?

9prady9 commented:
> #370 is in, can you please resolve merge conflicts?

Sure, will do as early as I can, at max couple of days. Thanks for the update.

9prady9 committed:
Fix cmake config file generation to use correct function

Earlier to this change, the codebase was using regular configure_file
to generate the package config file which renders the install files
non-relocatable. This commit refactors the regular configure_file to
configure_package_config_file command.

Also, fixed the include path values passed to hipSYCL-rt target

Refactor variable names and options to be consistent

9prady9 commented:
Sample cmake error output for examples when platform target is not provided
```bash
CMake Error at /home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/lib/cmake/hipSYCL/hipsycl-config.cmake:93 (message):
  More than one hipSYCL target platform is available.

  Must specify desired target(s) [and associated device(s)] via
  HIPSYCL_TARGETS, in the format
  HIPSYCL_TARGETS=<platform1>[:arch1[,arch2][..,archN]][;<platform2>[:arch1][...]][..;<platformN>]
  where platforms are one of cpu | cuda.
Call Stack (most recent call first):
  CMakeLists.txt:5 (find_package)
```

psath commented:
>   `where platforms are one of cpu | cuda.`

Minor, but with #370, this should be `omp | cuda`, `--hipsycl-targets` doesn't accept `cpu`

9prady9 commented:
> > `where platforms are one of cpu | cuda.`
> 
> Minor, but with #370, this should be `omp | cuda`, `--hipsycl-targets` doesn't accept `cpu`

Sorry, didn't notice that. Hm, it is now tricky, the variable now has CPU infix but the platform name in `omp`.  

Actually, even rocm backend is `hip` now. Have you guys considered renaming the corresponding cmake variables to reflect the same ?

Hm, there is another confusing thing:
- platform is set to `hip` in https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L42
- when checking for canonical names , it is set to `rocm` - https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L96

Which one is the correct name now ?

psath commented:
> > > `where platforms are one of cpu | cuda.`
> > 
> > 
> > Minor, but with #370, this should be `omp | cuda`, `--hipsycl-targets` doesn't accept `cpu`
> 
> Sorry, didn't notice that. Hm, it is now tricky, the variable now has CPU infix but the platform name in `omp`.
> 
> Actually, even rocm backend is `hip` now. Have you guys considered renaming the corresponding cmake variables to reflect the same ?
> 
> Hm, there is another confusing thing:
> 
> * platform is set to `hip` in https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L42
> * when checking for canonical names , it is set to `rocm` - https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L96
> 
> Which one is the correct name now ?

`--hipsycl-targets` only accepts `cuda | hip | omp`, it goes directly to `config.targets`, which in places like https://github.com/9prady9/hipSYCL/blob/cmake_config_fixes/bin/syclcc-clang#L665, only accepts that trio
`--hipsycl-platforms` (which is on its way out), accepts those three, plus many synonyms. https://github.com/9prady9/hipSYCL/blob/cmake_config_fixes/bin/syclcc-clang#L390
Based on @illuhad's comments in #370, I think we're trying to standardize to just the three supported by `targets`

IMO, standardizing the CMake names to match makes sense

illuhad commented:
> Based on @illuhad's comments in #370, I think we're trying to standardize to just the three supported by targets

Yes. Those synonyms have caused nothing but difficulties.

> IMO, standardizing the CMake names to match makes sense

Agreed.

> Minor, but with #370, this should be omp | cuda, --hipsycl-targets doesn't accept cpu

Yep. The motivation is that we need make some room for other CPU backends :-)

9prady9 commented:
@illuhad @psath Please check the most recent commit - the names should be consistent now.

The folllowing is the error message now.
```
CMake Error at /home/pradeep/gitroot/ArrayFireWorkspace/hipSYCL/build/pkg/lib/cmake/hipSYCL/hipsycl-config.cmake:91 (message):
  More than one hipSYCL target platform is available.

  Must specify desired target(s) [and associated device(s)] via
  HIPSYCL_TARGETS, in the format
  HIPSYCL_TARGETS=<platform1>[:arch1[,arch2][..,archN]][;<platform2>[:arch1][...]][..;<platformN>]
  where platforms are one of omp | cuda.
Call Stack (most recent call first):
  CMakeLists.txt:5 (find_package)
```

psath commented:
Looks good to me

9prady9 commented:
Note that, I didn't change the `CPU` and `ROCM` infixes in `CMakeLists.txt` as that would go beyond scope of this particular fix.

9prady9 commented:
Oops, name change triggered some failure. Not able to reproduce this locally. @illuhad @psath Any hints for me ?

psath commented:
Looks like the test scripts are still using the platform/arch combo. The `HIPSYCL_PLATFORM_CANONICAL` checks on https://github.com/9prady9/hipSYCL/blob/cmake_config_fixes/cmake/hipsycl-config.cmake.in#L107, 116, and 123 need to refer to `hip` not `rocm` now that the canonical term has changed

9prady9 commented:
> Oof, build failures are due to a regression in #370. Looks like the test scripts are still using the platform/arch combo. The `HIPSYCL_PLATFORM_CANONICAL` checks on https://github.com/9prady9/hipSYCL/blob/cmake_config_fixes/cmake/hipsycl-config.cmake.in#L107, 116, and 123 need to refer to `hip` not `rocm`

Interesting, shouldn't that give me an error even in local build ?

psath commented:
> > Oof, build failures are due to a regression in #370. Looks like the test scripts are still using the platform/arch combo. The `HIPSYCL_PLATFORM_CANONICAL` checks on https://github.com/9prady9/hipSYCL/blob/cmake_config_fixes/cmake/hipsycl-config.cmake.in#L107, 116, and 123 need to refer to `hip` not `rocm`
> 
> Interesting, should that give me an error even in local build ?

If you're using `HIPSYCL_PLATFORMS` I'd think it would, setting it to `rocm` would canonicalize to `hip`, just like directly setting it, and neither should result in the `--hipsycl-platform` flag getting set correctly.

9prady9 commented:
```diff
diff --git a/cmake/hipsycl-config.cmake.in b/cmake/hipsycl-config.cmake.in
index 41cebfb..5703b26 100644
--- a/cmake/hipsycl-config.cmake.in
+++ b/cmake/hipsycl-config.cmake.in
@@ -104,7 +104,7 @@ endif()
 
 set(HIPSYCL_ROCM_PATH "" CACHE STRING "The path to the ROCm installation directory.")
 if(HIPSYCL_ROCM_PATH)
-  if((HIPSYCL_PLATFORM_CANONICAL STREQUAL "rocm") OR (HIPSYCL_TARGETS MATCHES "hip"))
+  if((HIPSYCL_PLATFORM_CANONICAL STREQUAL "hip") OR (HIPSYCL_TARGETS MATCHES "hip"))
     set(HIPSYCL_SYCLCC_EXTRA_ARGS "${HIPSYCL_SYCLCC_EXTRA_ARGS} --hipsycl-rocm-path=${HIPSYCL_ROCM_PATH}")
   else()
     message(WARNING "HIPSYCL_ROCM_PATH (${HIPSYCL_ROCM_PATH}) is ignored for current platform (${HIPSYCL_PLATFORM})")
@@ -113,14 +113,14 @@ endif()
 
 set(HIPSYCL_GPU_ARCH "" CACHE STRING "(DEPRECATED. Use HIPSYCL_TARGETS) GPU architecture used by ROCm / CUDA.")
 if(HIPSYCL_GPU_ARCH AND NOT HIPSYCL_TARGETS)
-  if(HIPSYCL_PLATFORM_CANONICAL STREQUAL "cuda" OR HIPSYCL_PLATFORM_CANONICAL STREQUAL "rocm")
+  if(HIPSYCL_PLATFORM_CANONICAL STREQUAL "cuda" OR HIPSYCL_PLATFORM_CANONICAL STREQUAL "hip")
     set(HIPSYCL_SYCLCC_EXTRA_ARGS "${HIPSYCL_SYCLCC_EXTRA_ARGS} --hipsycl-gpu-arch=${HIPSYCL_GPU_ARCH}")
   else()
     message(WARNING "HIPSYCL_GPU_ARCH (${HIPSYCL_GPU_ARCH}) is ignored for current platform (${HIPSYCL_PLATFORM})")
   endif()
 else()
   set(_TMP $ENV{HIPSYCL_GPU_ARCH})
-  if((HIPSYCL_PLATFORM_CANONICAL STREQUAL "cuda" OR HIPSYCL_PLATFORM_CANONICAL STREQUAL "rocm") AND NOT _TMP)
+  if((HIPSYCL_PLATFORM_CANONICAL STREQUAL "cuda" OR HIPSYCL_PLATFORM_CANONICAL STREQUAL "hip") AND NOT _TMP)
     message(SEND_ERROR "Please specify HIPSYCL_GPU_ARCH")
   endif()
   unset(_TMP)
```
that should fix it right ? @psath

psath commented:
> that should fix it right ? @psath

I believe so, the clang10 / ubuntu20.04, CUDA 10 build seems to be failing on the 2nd of the 3 `if`s

9prady9 committed:
Fix canonical platform checks in cmake config script

9prady9 commented:
@psath tests are green

illuhad commented:
@hipsycl-ci-bot test this as well please.

illuhad commented:
In it goes, thank you very much!

9prady9 commented:
cool @psath @illuhad Thanks guys
